RandLAPACK/comps/rl_preconditioners.hh

```diff
@@ -0,0 +1,208 @@
+ *      can easily be solved by CG without preconditioning.
+ * 
+ * The input matrix A must be contiguous; no "LDA" is allowed.
+ * 
+ * On exit, the preconditioner with mu=0 would be defined by
+ * taking the columns of V_sk and scaling them by 1/sigma_sk.
+ * We do not perform this scaling ourselves, since other values
+ * of mu might be intended.
+ * 
+ * @param[in] layout_A
+ *      Either blas::Layout::RowMajor or blas::Layout::ColMajor.
+ * @param[in] m
+ *      The number of rows in A
+ * @param[in] n
+ *      The number of columns in A (and rows in M)
+ * @param[in] d
+ *      The number of rows in the sketched data matrix; must be >= n.
+ * @param[in] k
+ *      The number of nonzeros per column in the sketching operator
+ *      A common value is k=8 when n is on the order of a couple thousand.
+ * @param[in] A
+ *      A buffer for an m \times n matrix A. Interpreted in 
+ *      "layout_A" order.
+ * @param[out] V_sk
+ *      We require that V_sk.size() >= d*n. Letting "r" denote the return
+ *      value of this function, on exit the orthogonal factor of the preconditioner
+ *      is identified with the first n*r entries in V_sk.data(),
+ *      read in column-major format.
+ * @param[in] sigma_sk
+ *      We require sigma_sk.size() >= n.
+ * @param[in] threads
+ *      The number of OpenMP threads used during sketching.
+ * @param[in] seed_key
+ *      The key provided to an underlying counter-based random number generator.
+ * @param[inout] seed_ctr
+ *      A 32-bit counter provided to the underlying counter-based random number
+ *      generator. This value is incremented by m*k on-exit.
+ * @returns
+ *      void
+ */
+template <typename T>
+void rpc_data_svd_saso(
+    blas::Layout layout_A,
+    int64_t m, // number of rows in A
+    int64_t n, // number of columns in A
+    int64_t d, // number of rows in sketch of A
+    int64_t k, // number of nonzeros in each column of the sketching operator
+    std::vector<T>& A, // buffer of size m*n.
+    std::vector<T>& V_sk, // length at least d*n 
+    std::vector<T>& sigma_sk, // length at least n
+    int64_t threads, // number of OpenMP threads to use in sketching.
+    uint64_t seed_key,
+    uint32_t seed_ctr
+) {
+    T* buff_A = A.data();
+    if (V_sk.size() < d*n)
+        throw std::invalid_argument("V_sk must be of size at least d*n.");
+    T* buff_A_sk = V_sk.data();
+    
+    // step 1.1: define the sketching operator
+    RandBLAS::sparse::SparseDist D{
+        .n_rows = d,
+        .n_cols = m,
+        .vec_nnz=k
+    };
+    auto state = RandBLAS::base::RNGState(seed_key, seed_ctr);
+    RandBLAS::sparse::SparseSkOp<T> S(D, state);
+    RandBLAS::sparse::fill_sparse(S);
+
+    // step 1.2: sketch the data matrix
+    int64_t lda, lda_sk;
+    if (layout_A == blas::Layout::RowMajor) {
+        lda = n;
+        lda_sk = n;
+    } else {
+        lda = m;
+        lda_sk = d;
+    }
+    blas::scal(d*n, 0.0, buff_A_sk, 1);
```

Comment:
Question, if I initialize  the buffer to/get unlucky and allocate a page of all NaN before execution, what happens?

Unlike other operations in blaspp, the docs for blas::scal do not disavow that the impl will read the data when alpha = 0

Reply:
Thinking about this more; in single precision arithemetic there are 2**23 NaNs, so if we drew the initial value of the buffer from a uniform distribution over bit patterns there is in expectation d * n/512 NaNs. Which since NaN * 0 = NaN, means that this buffer for any reasonable sized matrix probably includes at least one NaN if the source of the initial condition is random. If we're dependent on this buffer being 0, I don't think this works.

Reply:
Fixed! Good catch.